### PR TITLE
[BEAM-1554] Add verification of Coder structural equality after encoding/decoding

### PR DESCRIPTION
**Pull request by iemejia**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by iemejia** [comment]
R: @kennknowles

**Comment by jbonofre** [comment]
No Jira ?

**Comment by iemejia** [comment]
I didn't think it was needed, since this is a minor change and I wanted also to see if Kenn agreed. But I will create one if it is needed.

**Comment by jbonofre** [comment]
No that's OK. Just wanted to be sure it's not a forget.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java [hunk 1]
```diff
@@ -230,26 +249,62 @@
...
-
     for (Coder.Context context : ALL_CONTEXTS) {
       CoderProperties.<T>structuralValueConsistentWithEqualsInContext(
           coder, context, value1, value2);
     }
   }
 
+  /**
+   * Verifies that for the given {@code Coder<T>}, {@code Coder.Context}, and
+   * values of type {@code T}, both the structural values are equal and the
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java]: In this case I think it verifies that `the structural values are equal if and only if the encoded bytes are equal`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java [hunk 2]
```diff
@@ -230,26 +249,62 @@
...
-
     assertEquals(
         coder.structuralValue(value1).equals(coder.structuralValue(value2)),
         Arrays.equals(
             encode(coder, context, value1),
             encode(coder, context, value2)));
   }
 
+  /**
+   * Verifies that for the given {@code Coder<T>}, {@code Coder.Context}, and
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java]: This method doesn't actually take a `Context`.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java [hunk 3]
```diff
@@ -230,26 +249,62 @@
...
+    }
+  }
+
+  /**
+   * Verifies that for the given {@code Coder<T>}, {@code Coder.Context}, and
+   * values of type {@code T}, the structural values are equal after encoding
+   * and decoding the values.
+   */
+  public static <T> void structuralValueDecodeEncodeEqualsInContext(
+          Coder<T> coder, Coder.Context context, T value1, T value2) throws Exception {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java]: I think what you want is `structuralValue(a) == structuralValue(decode(encode(a))`. This is a proxy for `a == decode(encode(a))` for things that don't have good `equals` implementations.

**Comment by kennknowles** [comment]
Since this is actually maybe useful for advanced users writing Coders, can you file a JIRA so it makes it automatically into the release notes?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java (hunk 1 above)
Comment (iemejia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java]: ok

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java (hunk 3 above)
Comment (iemejia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java]: Yes, you are absolutely right, that looks like a better test.

Comment (iemejia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java]: ok

**Comment by iemejia** [comment]
Ok, fixes done and rebased.

**Comment by kennknowles** [comment]
Looks great, I'll merge it. (w/ testing)

**Pull request merged**